File: backend/app/api/routes/images.py
import httpx
from fastapi import APIRouter, HTTPException

from app.schemas.images import GeneratedImage, ImageGenerationRequest, ImageGenerationResponse


router = APIRouter(prefix="/images", tags=["images"])


@router.post("/generate", response_model=ImageGenerationResponse)
async def generate_image(request: ImageGenerationRequest) -> ImageGenerationResponse:
    provider = request.provider.strip().lower()
    if provider not in {"openai", "custom"}:
        raise HTTPException(status_code=501, detail=f"Image provider '{request.provider}' is not supported yet.")

    api_key = (request.api_key or "").strip()
    if not api_key:
        raise HTTPException(status_code=400, detail="Image generation requires an API key.")

    base_url = (request.base_url or "https://api.openai.com/v1").strip().rstrip("/")
    endpoint = f"{base_url}/images/generations"
    payload = {
        "model": request.model,
        "prompt": request.prompt,
        "size": request.size,
    }

    try:
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                endpoint,
                headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
                json=payload,
            )
    except httpx.HTTPError as exc:
        raise HTTPException(status_code=502, detail=f"Image provider request failed: {exc}") from exc

    if response.status_code >= 400:
        raise HTTPException(status_code=502, detail=provider_error_detail(response))

    try:
        data = response.json()
    except ValueError as exc:
        raise HTTPException(status_code=502, detail="Image provider returned a non-JSON response.") from exc

    images = [parse_generated_image(item) for item in data.get("data", []) if isinstance(item, dict)]
    if not images:
        raise HTTPException(status_code=502, detail="Image provider returned no images.")

    return ImageGenerationResponse(
        status="success",
        message=f"Generated {len(images)} image{'s' if len(images) != 1 else ''}.",
        images=images,
        prompt=request.prompt,
    )


def parse_generated_image(item: dict) -> GeneratedImage:
    return GeneratedImage(
        url=item.get("url"),
        b64_json=item.get("b64_json"),
        revised_prompt=item.get("revised_prompt"),
    )


def provider_error_detail(response: httpx.Response) -> str:
    try:
        data = response.json()
    except ValueError:
        return response.text or f"Image provider returned HTTP {response.status_code}."

    if isinstance(data, dict):
        error = data.get("error")
        if isinstance(error, dict):
            message = error.get("message")
            if message:
                return str(message)
        detail = data.get("detail") or data.get("message")
        if detail:
            return str(detail)
    return f"Image provider returned HTTP {response.status_code}."

File: backend/internal/images/images.go
package images

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"

	"4ever/backend/internal/httputil"
	"github.com/gin-gonic/gin"
)

type Handler struct{}

type GenerationRequest struct {
	Provider string  `json:"provider" binding:"required"`
	BaseURL  *string `json:"base_url"`
	APIKey   *string `json:"api_key"`
	Model    string  `json:"model" binding:"required"`
	Prompt   string  `json:"prompt" binding:"required"`
	Size     string  `json:"size" binding:"required"`
}

type GeneratedImage struct {
	URL           any `json:"url,omitempty"`
	B64JSON       any `json:"b64_json,omitempty"`
	RevisedPrompt any `json:"revised_prompt,omitempty"`
}

type GenerationResponse struct {
	Status  string           `json:"status"`
	Message string           `json:"message"`
	Images  []GeneratedImage `json:"images"`
	Prompt  string           `json:"prompt"`
}

func Register(group *gin.RouterGroup, h Handler) {
	r := group.Group("/images")
	r.POST("/generate", h.Generate)
}

func (Handler) Generate(c *gin.Context) {
	var req GenerationRequest
	if !httputil.BindJSON(c, &req) {
		return
	}
	provider := strings.ToLower(strings.TrimSpace(req.Provider))
	if provider != "openai" && provider != "custom" {
		httputil.Error(c, http.StatusNotImplemented, "Image provider '"+req.Provider+"' is not supported yet.")
		return
	}
	apiKey := ""
	if req.APIKey != nil {
		apiKey = strings.TrimSpace(*req.APIKey)
	}
	if apiKey == "" {
		httputil.Error(c, http.StatusBadRequest, "Image generation requires an API key.")
		return
	}
	baseURL := "https://api.openai.com/v1"
	if req.BaseURL != nil && strings.TrimSpace(*req.BaseURL) != "" {
		baseURL = strings.TrimSpace(*req.BaseURL)
	}
	payload := map[string]any{"model": req.Model, "prompt": req.Prompt, "size": req.Size}
	body, _ := json.Marshal(payload)
	httpReq, _ := http.NewRequest(http.MethodPost, strings.TrimRight(baseURL, "/")+"/images/generations", bytes.NewReader(body))
	httpReq.Header.Set("Authorization", "Bearer "+apiKey)
	httpReq.Header.Set("Content-Type", "application/json")
	client := http.Client{Timeout: 120 * time.Second}
	resp, err := client.Do(httpReq)
	if err != nil {
		httputil.Error(c, http.StatusBadGateway, "Image provider request failed: "+err.Error())
		return
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		httputil.Error(c, http.StatusBadGateway, providerErrorDetail(respBody, resp.Status))
		return
	}
	var data map[string]any
	if err := json.Unmarshal(respBody, &data); err != nil {
		httputil.Error(c, http.StatusBadGateway, "Image provider returned a non-JSON response.")
		return
	}
	items, _ := data["data"].([]any)
	images := []GeneratedImage{}
	for _, item := range items {
		row, ok := item.(map[string]any)
		if !ok {
			continue
		}
		images = append(images, GeneratedImage{URL: row["url"], B64JSON: row["b64_json"], RevisedPrompt: row["revised_prompt"]})
	}
	if len(images) == 0 {
		httputil.Error(c, http.StatusBadGateway, "Image provider returned no images.")
		return
	}
	suffix := "s"
	if len(images) == 1 {
		suffix = ""
	}
	c.JSON(http.StatusOK, GenerationResponse{Status: "success", Message: "Generated " + strconvItoa(len(images)) + " image" + suffix + ".", Images: images, Prompt: req.Prompt})
}

func providerErrorDetail(body []byte, status string) string {
	var payload map[string]any
	if json.Unmarshal(body, &payload) == nil {
		if errorValue, ok := payload["error"].(map[string]any); ok {
			if message, ok := errorValue["message"].(string); ok && message != "" {
				return message
			}
		}
		for _, key := range []string{"detail", "message"} {
			if message, ok := payload[key].(string); ok && message != "" {
				return message
			}
		}
	}
	if len(body) > 0 {
		return string(body)
	}
	return "Image provider returned HTTP " + status + "."
}

func strconvItoa(value int) string {
	if value == 0 {
		return "0"
	}
	out := ""
	for value > 0 {
		out = string(rune('0'+value%10)) + out
		value /= 10
	}
	return out
}
